### PR TITLE
[CA-712] Sam tracing on APIs and Postgres/Ldap calls

### PR DESCRIPTION
**Pull request by andy7i**
Ticket: https://broadworkbench.atlassian.net/browse/CA-712

Adds Sam tracing on all APIs and Postgres/Ldap calls

example traces:
- multi-db trace: https://console.cloud.google.com/traces/traces?project=broad-dsde-dev&tid=4d594c5668d5107496f6ec7a6143afdc&pli=1&start=1586116332878&end=1588708332878&q=%2Fregister
- multi-call waterfall: https://console.cloud.google.com/traces/traces?project=broad-dsde-dev&tid=2e22becf7f399163b7170a741d253017&pli=1&start=1588576078442&end=1588619278442

Related firecloud-develop PR (to set trace sampling rate):
- https://github.com/broadinstitute/firecloud-develop/pull/2122

Some of the changes have been broken out into separate PRs. I'd be interested to hear if there are any ideas for splitting the PR further if that would be beneficial.

----

Change highlights:
- Akka and OpenCensus updates (merged in https://github.com/broadinstitute/sam/pull/428)
- Readme updates (reviewed in https://github.com/broadinstitute/sam/pull/432)
- Trace utils: in OpenCensusIOUtils
- `SamRequestContext` (created with the goal of making it easier to pass something to all functions of Sam without having to re-do all of the plumbing again)
- `withSamRequestContext` directive that has been added to the root of most routes (all routes minus swagger and status, which we don't want to trace)
- updates to LdapSupport and DatabaseSupport to add tracing to all postgres and ldap calls (that include a parentSpan)
- Pass the name of the postgres and ldap calls so that they can appear in the traces (see examples at the top)
- removes deprecated and unused code (could pull this out into a separate PR, but seems small enough)
- More plumbing than Super Mario World.
- Internal trace added for GoogleGroupSyncMonitor.

Note: Contexts without a parentSpan (`SamRequestContext(None)`) are used in the following cases:
- (most cases) in boot and init functions (where the function is called just once)
- the following are commented in: https://github.com/broadinstitute/sam/pull/421#pullrequestreview-406132463 : 
  - In `StandardUserInfoDirectives.scala` (see comment)

---

**PR checklist**

- [x] I've followed [the instructions](https://github.com/broadinstitute/sam/blob/develop/CONTRIBUTING.md#api-changes) if I've made any changes to the API, _especially_ if they're breaking changes
- [x] I've updated the RC_XXX release ticket with any manual steps required to release this change
- [x] I've updated the [FISMA documentation](https://github.com/broadinstitute/sam/blob/develop/CONTRIBUTING.md#fisma-documentation-changes) if I've made any security-related changes, including auth, encryption, or auditing

### CODE REVIEW

File: src/main/scala/org/broadinstitute/dsde/workbench/sam/util/TraceContext.scala [hunk 1]
```diff
@@ -0,0 +1,7 @@
+package org.broadinstitute.dsde.workbench.sam.util
+
+import io.opencensus.trace.Span
+
+// TraceContext was created with the goal of pre-empting a repeat of the crazy plumbing that was required to add the trace.
+// Instead of needing to add a new param to each API and underlying function, we can just add to this class instead.
+class TraceContext(val parentSpan: Span)
```
Comment (dvoet) [inline src/main/scala/org/broadinstitute/dsde/workbench/sam/util/TraceContext.scala]: I would call it `SamRequestContext` as future fields may have nothing to do with tracing. Also make it a case class.

**Comment by andy7i** [comment]
jenkins retest

**Comment by andy7i** [comment]
jenkins retest (fiab-start failure)

**Comment by andy7i** [comment]
jenkins retest - fiab start failure

File: README.md [hunk 2]
```diff
@@ -106,7 +106,7 @@ class SamClient(samBasePath: String) {
 
 Note that Sam does not actually launch workflows create VMs but appears to in this diagram in order to simplify interactions. The key concept is the user of service accounts.
 
-#### Google integration requires: 
+#### Google integration requires
```
Comment (andy7i) [inline README.md]: Readme changes are being reviewed separately: https://github.com/broadinstitute/sam/pull/432

File: automation/README.md [hunk 3]
```diff
@@ -9,35 +9,39 @@ See [firecloud-automated-testing](https://github.com/broadinstitute/firecloud-au
 
 ### Set Up
 
-Render configs:
+#### Render configs
```
Comment (andy7i) [inline automation/README.md]: likewise being reviewed in https://github.com/broadinstitute/sam/pull/432

File: src/main/scala/org/broadinstitute/dsde/workbench/sam/api/StandardUserInfoDirectives.scala [hunk 4]
```diff
@@ -97,33 +98,26 @@ object StandardUserInfoDirectives {
...
-      expiresInLong => getUserInfo(oidcHeaders.token, oidcHeaders.googleSubjectId, oidcHeaders.email, expiresInLong, directoryDAO)
+      expiresInLong => getUserInfo(oidcHeaders.token, oidcHeaders.googleSubjectId, oidcHeaders.email, expiresInLong, directoryDAO, samRequestContext)
     )
 
-  def getUserInfo(
-      token: OAuth2BearerToken,
-      googleSubjectId: GoogleSubjectId,
-      email: WorkbenchEmail,
-      expiresIn: Long,
-      directoryDAO: DirectoryDAO): IO[UserInfo] =
```
Comment (andy7i) [inline src/main/scala/org/broadinstitute/dsde/workbench/sam/api/StandardUserInfoDirectives.scala:113]: (auto-formatted by intellij)

File: src/main/scala/org/broadinstitute/dsde/workbench/sam/google/GoogleExtensions.scala [hunk 5]
```diff
@@ -260,64 +263,56 @@ class GoogleExtensions(
...
   override def onGroupDelete(groupEmail: WorkbenchEmail): Future[Unit] =
     googleDirectoryDAO.deleteGroup(groupEmail)
 
-  @deprecated("Use new two-argument version of this function", "Sam Phase 3")
-  def createUserPetServiceAccount(user: WorkbenchUser): Future[PetServiceAccount] =
-    createUserPetServiceAccount(user, petServiceAccountConfig.googleProject).unsafeToFuture()
-
-  @deprecated("Use new two-argument version of this function", "Sam Phase 3")
-  def deleteUserPetServiceAccount(userId: WorkbenchUserId): Future[Boolean] =
-    deleteUserPetServiceAccount(userId, petServiceAccountConfig.googleProject).unsafeToFuture() //TODO: shall we delete these deprecated methods
```
Comment (andy7i) [inline src/main/scala/org/broadinstitute/dsde/workbench/sam/google/GoogleExtensions.scala:302]: verified that they aren't used in the code elsewhere ; could remove this to be handled in a separate PR

File: src/main/scala/org/broadinstitute/dsde/workbench/sam/util/OpenCensusIOUtils.scala [hunk 6]
```diff
@@ -1,34 +1,53 @@
...
-                            parentSpan: Span,
-                            failureStatus: Throwable => Status = (_: Throwable) => Status.UNKNOWN
-                          )(f: Span => IO[T]): IO[T] =
-    traceIOSpan(IO(startSpanWithParent(name, parentSpan)), failureStatus)(f)
+  def traceIOWithContext[T](
+                             name: String,
+                             samRequestContext: SamRequestContext,
+                             failureStatus: Throwable => Status = (_: Throwable) => Status.UNKNOWN
+                          )(f: SamRequestContext => IO[T]): IO[T] = {
+    if (samRequestContext.parentSpan == null) { // ignore calls from unit tests and calls on boot; an alternative here would be to have a MockOpenCensusIOUtils to unit tests
```
Comment (andy7i) [inline src/main/scala/org/broadinstitute/dsde/workbench/sam/util/OpenCensusIOUtils.scala]: I'd be interested to know if anyone thinks we should use something like MockOpenCensusIOUtils for unit tests instead of checking for null

File: src/main/scala/org/broadinstitute/dsde/workbench/sam/service/StatusService.scala [hunk 7]
```diff
@@ -35,7 +36,7 @@ class StatusService(
 
   private def checkOpenDJ(groupToLoad: WorkbenchGroupName): IO[SubsystemStatus] = {
     logger.info("checking opendj connection")
-    directoryDAO.loadGroupEmail(groupToLoad).map {
+    directoryDAO.loadGroupEmail(groupToLoad, SamRequestContext(null)).map {
```
Comment (andy7i) [inline src/main/scala/org/broadinstitute/dsde/workbench/sam/service/StatusService.scala]: this is used from `healthMonitor` on line 29. Any issues with doing it like this?

File: src/main/scala/org/broadinstitute/dsde/workbench/sam/google/GoogleGroupSyncMonitor.scala [hunk 8]
```diff
@@ -130,7 +131,7 @@ class GoogleGroupSyncMonitorActor(
       val groupId: WorkbenchGroupIdentity = parseMessage(message)
 
       groupSynchronizer
-        .synchronizeGroupMembers(groupId)
+        .synchronizeGroupMembers(groupId, samRequestContext = SamRequestContext(null))
```
Comment (andy7i) [inline src/main/scala/org/broadinstitute/dsde/workbench/sam/google/GoogleGroupSyncMonitor.scala]: does the `GoogleGroupSyncMonitorActor` need to have tracing?

File: src/main/scala/org/broadinstitute/dsde/workbench/sam/api/StandardUserInfoDirectives.scala [hunk 9]
```diff
@@ -67,12 +68,12 @@ trait StandardUserInfoDirectives extends UserInfoDirectives with LazyLogging {
...
   private def requireUserInfoFromJwt(jwtUserInfo: JwtUserInfo, bearerToken: OAuth2BearerToken, icService: IdentityConcentratorService): Directive1[UserInfo] =
     onSuccess {
-      getUserInfoFromJwt(jwtUserInfo, bearerToken, directoryDAO, icService).unsafeToFuture()
+      getUserInfoFromJwt(jwtUserInfo, bearerToken, directoryDAO, icService, SamRequestContext(null)).unsafeToFuture()
     }
 
   private def requireUserInfoFromOIDC(oidcHeaders: OIDCHeaders): Directive1[UserInfo] =
     onSuccess {
-      getUserInfoFromOidcHeaders(directoryDAO, oidcHeaders).unsafeToFuture()
+      getUserInfoFromOidcHeaders(directoryDAO, oidcHeaders, SamRequestContext(null)).unsafeToFuture()
```
Comment (andy7i) [inline src/main/scala/org/broadinstitute/dsde/workbench/sam/api/StandardUserInfoDirectives.scala]: for `requireUserInfoFromJwt` and `requireUserInfoFromOIDC`, tracing seems unnecessary since they are used to authenticate the user?

File: src/main/scala/org/broadinstitute/dsde/workbench/sam/api/SamModelDirectives.scala [hunk 10]
```diff
@@ -7,15 +7,16 @@ import akka.http.scaladsl.unmarshalling.FromRequestUnmarshaller
...
 /**
   * Created by gpolumbo on 3/26/2018
   */
 trait SamModelDirectives {
   val userService: UserService
 
-  def withSubject(email: WorkbenchEmail): Directive1[WorkbenchSubject] =
-    onSuccess(userService.getSubjectFromEmail(email)).map {
+  def withSubject(email: WorkbenchEmail, samRequestContext: SamRequestContext = SamRequestContext(null)): Directive1[WorkbenchSubject] =
+    onSuccess(userService.getSubjectFromEmail(email, samRequestContext)).map {
```
Comment (andy7i) [inline src/main/scala/org/broadinstitute/dsde/workbench/sam/api/SamModelDirectives.scala]: Do we want this to be a part of the trace? Trying to add tracing by changing the order leads to some errors (which may be easily solved if this is the behavior we want)

when reordering the code:

from
```
      withSubject(WorkbenchEmail(email)) { subject =>
        completeWithTrace(samRequestContext =>
          managedGroupService.addSubjectToPolicy(managedGroup.resourceId, accessPolicyName, subject, samRequestContext).map(_ => StatusCodes.NoContent))
      }
```
to
```
        completeWithTrace{samRequestContext =>
          withSubject(WorkbenchEmail(email)) { subject =>
            managedGroupService.addSubjectToPolicy(managedGroup.resourceId, accessPolicyName, subject, samRequestContext).map(_ => StatusCodes.NoContent)
          }
        }
```
results in this error:
```
[error] /Users/ajang/Firecloud/sam/src/main/scala/org/broadinstitute/dsde/workbench/sam/api/ManagedGroupRoutes.scala:123:126: type mismatch;
[error]  found   : scala.concurrent.Future[akka.http.scaladsl.model.StatusCodes.Success]
[error]  required: akka.http.scaladsl.server.RequestContext => scala.concurrent.Future[akka.http.scaladsl.server.RouteResult]
[error]             managedGroupService.addSubjectToPolicy(managedGroup.resourceId, accessPolicyName, subject, samRequestContext).map(_ => StatusCodes.NoContent)
[error]                                                                                                                              ^
[error] one error found
[error] (Compile / compileIncremental) Compilation failed
```

File: src/main/scala/org/broadinstitute/dsde/workbench/sam/api/SecurityDirectives.scala [hunk 11]
```diff
@@ -10,23 +10,24 @@ import org.broadinstitute.dsde.workbench.sam.service.PolicyEvaluatorService
...
 
-  def requireAction(resource: FullyQualifiedResourceId, action: ResourceAction, userId: WorkbenchUserId): Directive0 =
+  def requireAction(resource: FullyQualifiedResourceId, action: ResourceAction, userId: WorkbenchUserId, samRequestContext: SamRequestContext = SamRequestContext(null)): Directive0 =
     requireOneOfAction(resource, Set(action), userId)
 
-  def requireOneOfAction(resource: FullyQualifiedResourceId, requestedActions: Set[ResourceAction], userId: WorkbenchUserId): Directive0 =
+  def requireOneOfAction(resource: FullyQualifiedResourceId, requestedActions: Set[ResourceAction], userId: WorkbenchUserId, samRequestContext: SamRequestContext = SamRequestContext(null)): Directive0 =
     Directives.mapInnerRoute { innerRoute =>
-      onSuccess(hasPermissionOneOf(resource, requestedActions, userId)) { hasPermission =>
+      onSuccess(hasPermissionOneOf(resource, requestedActions, userId, samRequestContext)) { hasPermission =>
```
Comment (andy7i) [inline src/main/scala/org/broadinstitute/dsde/workbench/sam/api/SecurityDirectives.scala]: Do we want this to be a part of the trace? Trying to add tracing by changing the order leads to some errors (which may be easily solved if this is the behavior we want)

when reordering the code:

from
```
    requireAction(managedGroup, SamResourceActions.readPolicy(accessPolicyName), userInfo.userId) {
      completeWithTrace(samRequestContext =>
        managedGroupService.listPolicyMemberEmails(managedGroup.resourceId, accessPolicyName, samRequestContext).map(x => StatusCodes.OK -> x.toSet))
    }
```
to
```
    completeWithTrace { samRequestContext =>
      requireAction(managedGroup, SamResourceActions.readPolicy(accessPolicyName), userInfo.userId) {
        managedGroupService.listPolicyMemberEmails(managedGroup.resourceId, accessPolicyName, samRequestContext).map(x => StatusCodes.OK -> x.toSet)
      }
    }
```
results in:
```
[error] /Users/ajang/Firecloud/sam/src/main/scala/org/broadinstitute/dsde/workbench/sam/api/ManagedGroupRoutes.scala:104:117: type mismatch;
[error]  found   : cats.effect.IO[(akka.http.scaladsl.model.StatusCodes.Success, scala.collection.immutable.Set[org.broadinstitute.dsde.workbench.model.WorkbenchEmail])]
[error]  required: akka.http.scaladsl.server.Route
[error]     (which expands to)  akka.http.scaladsl.server.RequestContext => scala.concurrent.Future[akka.http.scaladsl.server.RouteResult]
[error]         managedGroupService.listPolicyMemberEmails(managedGroup.resourceId, accessPolicyName, samRequestContext).map(x => StatusCodes.OK -> x.toSet)
[error]                                                                                                                     ^
[error] one error found
```

**Review by andy7i — COMMENTED** [review]
comments on SamRequestContext(null) usage

File: src/main/scala/org/broadinstitute/dsde/workbench/sam/google/GoogleGroupSyncMonitor.scala (hunk 8 above)
Comment (dvoet) [reply src/main/scala/org/broadinstitute/dsde/workbench/sam/google/GoogleGroupSyncMonitor.scala]: I think it would be good to have tracing here but could be another ticket and definitely a follow on PR.

File: src/main/scala/org/broadinstitute/dsde/workbench/sam/service/StatusService.scala (hunk 7 above)
Comment (dvoet) [reply src/main/scala/org/broadinstitute/dsde/workbench/sam/service/StatusService.scala]: this is fine

File: src/main/scala/org/broadinstitute/dsde/workbench/sam/api/SamModelDirectives.scala (hunk 10 above)
Comment (dvoet) [reply src/main/scala/org/broadinstitute/dsde/workbench/sam/api/SamModelDirectives.scala]: the problem with this and the `requireAction` stuff is that `withSubject` and `requireAction` are both `Directives` and `completeWithTrace` can't contain a `Directive`. 

I better approach would be instead of `completeWithTrace` would be to have a `withTrace` directive.

Comment (dvoet) [reply src/main/scala/org/broadinstitute/dsde/workbench/sam/api/SamModelDirectives.scala]:
```
def withSamRequestContext: Directive1[SamRequestContext] = traceRequest.map { span => SamRequestContext(span) }
```

**Comment by andy7i** [comment]
jenkins retest -- failed fiab-start

File: src/main/scala/org/broadinstitute/dsde/workbench/sam/api/SamModelDirectives.scala (hunk 10 above)
Comment (andy7i) [reply src/main/scala/org/broadinstitute/dsde/workbench/sam/api/SamModelDirectives.scala]: //implemented

File: src/main/scala/org/broadinstitute/dsde/workbench/sam/api/SecurityDirectives.scala (hunk 11 above)
Comment (andy7i) [reply src/main/scala/org/broadinstitute/dsde/workbench/sam/api/SecurityDirectives.scala]: //implemented

File: src/main/scala/org/broadinstitute/dsde/workbench/sam/google/GoogleGroupSyncMonitor.scala (hunk 8 above)
Comment (andy7i) [reply src/main/scala/org/broadinstitute/dsde/workbench/sam/google/GoogleGroupSyncMonitor.scala]: New ticket: https://broadworkbench.atlassian.net/browse/CA-817

Should I start working on this immediately or should it be prioritized in the backlog?

File: src/main/scala/org/broadinstitute/dsde/workbench/sam/google/GoogleGroupSyncMonitor.scala [hunk 12]
```diff
@@ -130,7 +131,7 @@ class GoogleGroupSyncMonitorActor(
       val groupId: WorkbenchGroupIdentity = parseMessage(message)
 
       groupSynchronizer
-        .synchronizeGroupMembers(groupId)
+        .synchronizeGroupMembers(groupId, samRequestContext = SamRequestContext(None))
```
Comment (wchamber) [inline src/main/scala/org/broadinstitute/dsde/workbench/sam/google/GoogleGroupSyncMonitor.scala]: WDYT about starting a trace here? It's an internal kind of request, but I think tracing it and annotating as such could be interesting.

File: src/main/scala/org/broadinstitute/dsde/workbench/sam/openam/PostgresAccessPolicyDAO.scala [hunk 13]
```diff
@@ -533,12 +533,12 @@ class PostgresAccessPolicyDAO(protected val dbRef: DbReference,
                and ${p.public} = true"""
 
     import SamTypeBinders._
-    runInTransaction { implicit session =>
+    runInTransaction("listPublicAccessPolicies", samRequestContext)({ implicit session =>
```
Comment (wchamber) [inline src/main/scala/org/broadinstitute/dsde/workbench/sam/openam/PostgresAccessPolicyDAO.scala]: can we differentiate the names of overloaded functions? There are multiple 'listPublicAccessPolicies' and we should be able to see them differently in traces. (Are there other overloaded functions?)

File: src/main/scala/org/broadinstitute/dsde/workbench/sam/service/PolicyEvaluatorService.scala [hunk 14]
```diff
@@ -30,34 +30,34 @@ class PolicyEvaluatorService(
           Set.empty,
           Set(SamResourceActions.setPublicPolicy(ManagedGroupService.adminNotifierPolicyName)),
           true
-        ))
+        ), SamRequestContext(None))
```
Comment (wchamber) [inline src/main/scala/org/broadinstitute/dsde/workbench/sam/service/PolicyEvaluatorService.scala]: ditto internal trace

File: src/main/scala/org/broadinstitute/dsde/workbench/sam/service/StatusService.scala [hunk 15]
```diff
@@ -35,7 +36,7 @@ class StatusService(
 
   private def checkOpenDJ(groupToLoad: WorkbenchGroupName): IO[SubsystemStatus] = {
     logger.info("checking opendj connection")
-    directoryDAO.loadGroupEmail(groupToLoad).map {
+    directoryDAO.loadGroupEmail(groupToLoad, SamRequestContext(None)).map {
```
Comment (wchamber) [inline src/main/scala/org/broadinstitute/dsde/workbench/sam/service/StatusService.scala]: can status checks have request contexts?

File: src/main/scala/org/broadinstitute/dsde/workbench/sam/util/DatabaseSupport.scala [hunk 16]
```diff
@@ -11,9 +12,12 @@ trait DatabaseSupport {
   protected val cs: ContextShift[IO]
   protected val dbRef: DbReference
 
-  protected def runInTransaction[A](databaseFunction: DBSession => A): IO[A] = {
-    cs.evalOn(ecForDatabaseIO)(IO {
-      dbRef.inLocalTransaction(databaseFunction)
-    })
+  protected def runInTransaction[A](dbQueryName: String, samRequestContext: SamRequestContext)(databaseFunction: DBSession => A): IO[A] = {
```
Comment (wchamber) [inline src/main/scala/org/broadinstitute/dsde/workbench/sam/util/DatabaseSupport.scala:22]: let's add a comment about what dbQueryName and samRequestContext values should be.

File: src/main/scala/org/broadinstitute/dsde/workbench/sam/util/LdapSupport.scala [hunk 17]
```diff
@@ -40,5 +41,7 @@ trait LdapSupport {
       email <- getAttribute(results, Attr.email).toRight(s"${Attr.email} attribute missing")
     } yield WorkbenchUser(WorkbenchUserId(uid), getAttribute(results, Attr.googleSubjectId).map(GoogleSubjectId), WorkbenchEmail(email), None)
 
-  protected def executeLdap[A](ioa: IO[A]): IO[A] = cs.evalOn(ecForLdapBlockingIO)(ioa)
+  protected def executeLdap[A](ioa: IO[A], dbQueryName: String, samRequestContext: SamRequestContext): IO[A] = {
```
Comment (wchamber) [inline src/main/scala/org/broadinstitute/dsde/workbench/sam/util/LdapSupport.scala:52]: ditto comment for variables

File: src/main/scala/org/broadinstitute/dsde/workbench/sam/util/SamRequestContext.scala [hunk 18]
```diff
@@ -0,0 +1,10 @@
+package org.broadinstitute.dsde.workbench.sam.util
+
+import io.opencensus.trace.Span
+
+/*
+SamRequestContext was created with the goal of pre-empting a repeat of the crazy plumbing that was required to add
+the trace. In the future, instead of needing to add a new param to each API and underlying function, we can just add
+to this class instead.
+ */
+case class SamRequestContext (parentSpan: Option[Span])
```
Comment (wchamber) [inline src/main/scala/org/broadinstitute/dsde/workbench/sam/util/SamRequestContext.scala:16]: I don't know how Scala case classes work, but could we have a factory function for creating the SamRequestContext from a Span and one for from None? I think passing None is not really a "normal" use, and having a factory function to do it can represent that.

SamRequestContext contextForTest() {...}

**Review by wchamber — CHANGES_REQUESTED** [review]
I'm not wild about not having a trace for the internal calls. I think we should try, if possible, to reserve the SamRequestContext(None) initialization for testing.
(can the PR be marked as not a draft in github?)

File: src/main/scala/org/broadinstitute/dsde/workbench/sam/google/GoogleGroupSyncMonitor.scala (hunk 8 above)
Comment (wchamber) [reply src/main/scala/org/broadinstitute/dsde/workbench/sam/google/GoogleGroupSyncMonitor.scala]: I think it's like 3 lines of code, but we can separate it.
https://opencensus.io/quickstart/java/tracing/#using-the-tracer

File: src/main/scala/org/broadinstitute/dsde/workbench/sam/api/StandardUserInfoDirectives.scala (hunk 9 above)
Comment (wchamber) [reply src/main/scala/org/broadinstitute/dsde/workbench/sam/api/StandardUserInfoDirectives.scala]: Still part of the request, so we still care how long it takes would be my opinion. Unless it's an absolute pain to add due to how we have our akka structured.
(missed this on the first go around.)

File: src/main/scala/org/broadinstitute/dsde/workbench/sam/util/OpenCensusIOUtils.scala (hunk 6 above)
Comment (wchamber) [reply src/main/scala/org/broadinstitute/dsde/workbench/sam/util/OpenCensusIOUtils.scala]: why is it null instead of None?
I don't think we care about asserting anything about the traces, but if it's painless to use a fake trace, that's better than None since the fake trace is "closer" to reality for testing adding spans. But if it's much work at all, not worth it imo.

File: src/main/scala/org/broadinstitute/dsde/workbench/sam/google/GoogleGroupSyncMonitor.scala (hunk 12 above)
Comment (andy7i) [reply src/main/scala/org/broadinstitute/dsde/workbench/sam/google/GoogleGroupSyncMonitor.scala]: I think it would be pretty easy to start a trace here. Would we want to start tracing higher up? That might be a bit more difficult, and doug was suggesting pulling this out into a new PR in this comment on "outdated" code: https://github.com/broadinstitute/sam/pull/421#discussion_r420398113

File: src/main/scala/org/broadinstitute/dsde/workbench/sam/openam/PostgresAccessPolicyDAO.scala (hunk 13 above)
Comment (dvoet) [reply src/main/scala/org/broadinstitute/dsde/workbench/sam/openam/PostgresAccessPolicyDAO.scala]: we can un-overload them if need be

File: src/main/scala/org/broadinstitute/dsde/workbench/sam/util/SamRequestContext.scala (hunk 18 above)
Comment (dvoet) [reply src/main/scala/org/broadinstitute/dsde/workbench/sam/util/SamRequestContext.scala:16]: I think there are certainly places where we don't want tracing. The status api call comes to mind because they will only generate noise (something like 80% of our api calls are to the status api which means we would have to turn up our sampling rate in order to get a reasonable sample of the other api calls.

File: src/main/scala/org/broadinstitute/dsde/workbench/sam/api/StandardUserInfoDirectives.scala [hunk 19]
```diff
@@ -67,12 +68,12 @@ trait StandardUserInfoDirectives extends UserInfoDirectives with LazyLogging {
 
   private def requireUserInfoFromJwt(jwtUserInfo: JwtUserInfo, bearerToken: OAuth2BearerToken, icService: IdentityConcentratorService): Directive1[UserInfo] =
     onSuccess {
-      getUserInfoFromJwt(jwtUserInfo, bearerToken, directoryDAO, icService).unsafeToFuture()
+      getUserInfoFromJwt(jwtUserInfo, bearerToken, directoryDAO, icService, SamRequestContext(None)).unsafeToFuture()
```
Comment (dvoet) [inline src/main/scala/org/broadinstitute/dsde/workbench/sam/api/StandardUserInfoDirectives.scala]: I would like to see the `requireUserInfo` and `requireCreateUser` functions traced. If you would like to open a follow on jira that is fine.

File: src/main/scala/org/broadinstitute/dsde/workbench/sam/service/ResourceService.scala [hunk 20]
```diff
@@ -47,22 +47,22 @@ class ResourceService(
...
-  def initResourceTypes(): IO[Iterable[ResourceType]] =
+  def initResourceTypes(samRequestContext: SamRequestContext = SamRequestContext(None)): IO[Iterable[ResourceType]] =
     resourceTypes.get(SamResourceTypes.resourceTypeAdminName) match {
       case None =>
         IO.raiseError(new WorkbenchException(s"Could not initialize resource types because ${SamResourceTypes.resourceTypeAdminName.value} does not exist."))
       case Some(resourceTypeAdmin) =>
         for {
           // make sure resource type admin is added first because the rest depends on it
-          createdAdminType <- createResourceType(resourceTypeAdmin)
+          createdAdminType <- createResourceType(resourceTypeAdmin, SamRequestContext(None))
```
Comment (dvoet) [inline src/main/scala/org/broadinstitute/dsde/workbench/sam/service/ResourceService.scala]: use `samRequestContext` within this function

File: src/main/scala/org/broadinstitute/dsde/workbench/sam/util/DatabaseSupport.scala [hunk 21]
```diff
@@ -11,9 +12,12 @@ trait DatabaseSupport {
...
+    val spanName = "postgres-" + dbQueryName
+    traceIOWithContext(spanName, samRequestContext) { _ =>
```
Comment (dvoet) [inline src/main/scala/org/broadinstitute/dsde/workbench/sam/util/DatabaseSupport.scala]: I think this should be after the `cs.evalOn` call. The way it currently is, the span includes time to wait for the context switch.

File: src/main/scala/org/broadinstitute/dsde/workbench/sam/util/LdapSupport.scala [hunk 22]
```diff
@@ -40,5 +41,7 @@ trait LdapSupport {
...
+    traceIOWithContext("ldap-" + dbQueryName, samRequestContext)(_ => cs.evalOn(ecForLdapBlockingIO)(ioa))
```
Comment (dvoet) [inline src/main/scala/org/broadinstitute/dsde/workbench/sam/util/LdapSupport.scala]: same comment as in `runInTransaction`, trace should be after the context switch

**Review by dvoet — CHANGES_REQUESTED** [review]

File: src/main/scala/org/broadinstitute/dsde/workbench/sam/openam/PostgresAccessPolicyDAO.scala (hunk 13 above)
Comment (andy7i) [reply src/main/scala/org/broadinstitute/dsde/workbench/sam/openam/PostgresAccessPolicyDAO.scala]: Good catch -- I didn't notice that we had an overloaded function here. I can either rename/un-overload the functions or change the name that we pass to the trace. Any reason to go through the additional effort of the former over the latter?

File: src/main/scala/org/broadinstitute/dsde/workbench/sam/service/PolicyEvaluatorService.scala (hunk 14 above)
Comment (andy7i) [reply src/main/scala/org/broadinstitute/dsde/workbench/sam/service/PolicyEvaluatorService.scala]: Intellij says this is only called on Boot. Do you think this call (and the other similar 1-time init/boot calls) are worth tracing as well? I was leaning toward no but am curious to know other thoughts.

File: src/main/scala/org/broadinstitute/dsde/workbench/sam/service/StatusService.scala (hunk 15 above)
Comment (andy7i) [reply src/main/scala/org/broadinstitute/dsde/workbench/sam/service/StatusService.scala]: I could add that in (with no trace) if that's something we want. After discussing with Doug, the Status and Swagger routes were left without traces, but we could add in the plumbing for the context to one or both.

File: src/main/scala/org/broadinstitute/dsde/workbench/sam/util/SamRequestContext.scala (hunk 18 above)
Comment (andy7i) [reply src/main/scala/org/broadinstitute/dsde/workbench/sam/util/SamRequestContext.scala:16]: I also thought it would be nice to have a factory function for contexts with no traces. I have a shelved branch toward that goal. It was turning into a more-than-trivial amount of work and doug recommended that we use Option[Span] instead. I can reopen that if we want to go that route?

File: src/main/scala/org/broadinstitute/dsde/workbench/sam/google/GoogleGroupSyncMonitor.scala (hunk 12 above)
Comment (wchamber) [reply src/main/scala/org/broadinstitute/dsde/workbench/sam/google/GoogleGroupSyncMonitor.scala]: New PR/ticket sounds fine. Sorry for ballooning the pr scope :)
We might want it higher up wherever the pub/sub message first gets grabbed from the queue.

File: src/main/scala/org/broadinstitute/dsde/workbench/sam/openam/PostgresAccessPolicyDAO.scala (hunk 13 above)
Comment (wchamber) [reply src/main/scala/org/broadinstitute/dsde/workbench/sam/openam/PostgresAccessPolicyDAO.scala]: just changing the trace name sounds fine to me and I don't think we're trying to forbid overloads.

File: src/main/scala/org/broadinstitute/dsde/workbench/sam/service/PolicyEvaluatorService.scala (hunk 14 above)
Comment (wchamber) [reply src/main/scala/org/broadinstitute/dsde/workbench/sam/service/PolicyEvaluatorService.scala]: Probably not super interesting to trace boot functions (though would be good if we try to optimize start up time), so seems fine to leave for later.
The only reason to care more deeply is if they seem likely to be used additional times.
Let's just leave a comment about why we're not bothering to trace it for now.

File: src/main/scala/org/broadinstitute/dsde/workbench/sam/util/SamRequestContext.scala (hunk 18 above)
Comment (wchamber) [reply src/main/scala/org/broadinstitute/dsde/workbench/sam/util/SamRequestContext.scala:16]: Option[Span] is ok, but let's comment on why we might not have a Span.
If there are multiple reasons, a factory function is harder to name. What might get interesting is when/if we add a second variable to SamRequestContext. Then we might want to restrict how it's built. But if it's more-than-trivial, we can wait until then and eat the work later.

If we're sampling with a probability, whether we're sampling the status API doesn't affect any other API. All calls have the same likelihood of being sampled; they're all independent. Sampling status will just cause us to have more traces, so just cost us more to store. So if we don't want to do that, sounds ok.

**Comment by andy7i** [comment]
jenkins retest - fiab-start errors when starting consent, agora, bond

File: src/main/scala/org/broadinstitute/dsde/workbench/sam/api/StandardUserInfoDirectives.scala (hunk 19 above)
Comment (andy7i) [reply src/main/scala/org/broadinstitute/dsde/workbench/sam/api/StandardUserInfoDirectives.scala]: do you want them individually traced (with new spans for each) or is it sufficient to move the trace to start before requireUserInfo and requireCreateUser?

File: src/main/scala/org/broadinstitute/dsde/workbench/sam/google/GoogleGroupSyncMonitor.scala (hunk 12 above)
Comment (andy7i) [reply src/main/scala/org/broadinstitute/dsde/workbench/sam/google/GoogleGroupSyncMonitor.scala]: No worries! I started a trace at the beginning of this case statement. Do we need a trace for the case ReportMessage?

File: src/main/scala/org/broadinstitute/dsde/workbench/sam/api/StandardUserInfoDirectives.scala (hunk 9 above)
Comment (andy7i) [reply src/main/scala/org/broadinstitute/dsde/workbench/sam/api/StandardUserInfoDirectives.scala]: as discussed in standup, we'll pull this into a separate ticket/PR: https://broadworkbench.atlassian.net/browse/CA-849

**Review by wchamber — APPROVED** [review]

**Review by dvoet — APPROVED** [review]
